Fuel moisture model

janmandel opened the pull request:
TYPE: new feature/enhancement

KEYWORDS: WRF-Fire, fuel moisture, coupled model, equilibrium

SOURCE: Jan Mandel, Angel Farguell (University of Colorado Denver), Adam Kochanski (University of Utah)

DESCRIPTION OF CHANGES:   
Fuel moisture content (FMC) is now maintained in each fire model mesh cell separately; it was a single constraint for the whole model before. A FMC model is coupled with the coupled fire-atmosphere model. FMC has a strong effect on fire spread, including the diurnal cycle, and previously was currently a constant in the fuel description. The FMC model runs a differential equation for the FMC in each fuel class (10h, 100h, 1000h) in each atmospheric mesh cell. The FMC evolves towards an equilibrium, computed from the atmospheric state, with a speed dependent on the fuel class. The FMC is interpolated to the fire mesh and mixed in proportions given in fuel description to obtain the FMC on the fire mesh, which  is then used by the fire model. The code also supports import of the FMC map on the fire mesh. Two FMC test cases were added.

Reference: J. Mandel, S. Amram, J.D. Beezley, G. Kelman, A.K. Kochanski, V.Y. Kondratenko, B.H. Lynn, B. Regev, M. Vejmelka, Recent advances and applications of WRF–SFIRE. Natural Hazards and Earth System Science, 14, 829-2845, 2014, https://doi.org/10.5194/nhess-14-2829-2014

**An ideal test problem added**

The ideal example in test/em_fire/rain was added, which will run coupled fuel moisture and fire. The directory and the soft links there are created by ./compile em_fire and deleted by ./clean -a. Variables required to initialize surface physics are set in ideal.exe when built by ./compile em_fire.

**Flags required to be set in namelist.input**

See the file test/em_fire/namelist.input_rain for an example with annotations.

**sf_sfclay_physics**         in &physics section
A nonzero value to turn on a surface physics model is required in order for the variables T2, Q2, PSFC to be populated with values, which are needed to drive the FMC model. Other appropriate surface physics flags and input data are up to the user. The ideal.exe generated by compile em_fire sets the appropriate arrays for
**sf_sfclay_physics              =1**
**sf_surface_physics           =1**
which are therefore required in an ideal run if the fuel moisture model is turned on.

**Flags added to file namelist.input**

The following variables were added  to the &fire section:
**fmoist_run**  default    .false. :          Run the fuel FMC model (on the atmospheric grid).
**fmoist_interp** default .false.  :         Interpolate and mix FMC for the actual fuels on the fire grid.
**fmoist_only**  default  .false. :           Only run FMC model, do not run the fire model.
**fmoist_freq**     default   0 :              frequency to run the FMC model. 0: use fmoist_dt, >0: every fmoist_freq timesteps
**fmoist_dt**     default    600 :            FMC model time step (s).
**fire_fmc_read**  default  1 :               How to set FMC map on the fire mesh:  0: read on start as  array fmc_g in wrfinput;  required when the FMC model is run. 1: set initially as constant value from namelist.fire.  2: read from a text file (in ideal.exe only).

**If the flags above are not set and left at default, the FMC model does not run and there is no change in results. To turn the FMC model model on, it is recommended to specify fmoist_run=.true., fmoist_interp=.true., fire_fmc_read=0, and leave the rest of new flags at default**

**Variables added to file namelist.fire:**

Fuel properties **fmc_gw01** to **fmc_gw05** were added to the fuel descriptions in  the **&fuel_categories** section. Each variable **fmc_gw0**x contains the proportion of the fuel that is in moisture class x, for all fuel categories. For each fuel category, the proportions over the moisture classes should add to one. The default values used are from Albini, 1976 and are programmed in phys/module_fr_fire_phys.F.  See  test/em_fire/namelist.fire_fmc for an annotated example how to set the values instead of using the default.

New section **&moisture** was added, which allows to set the following parameters of the fuel moisture model. See  test/em_fire/namelist.fire_fmc for an annotated example, which contains the same values as also built into WRF. These parameter values were calibrated to the FMC model in the Canadian Fire Danger Rating System, for 10-hour fuel and scaled to other moisture classes. The &moisture section may be empty, but it must be present.

**moisture_classes**   number of moisture classes (max. 5)
**moisture_class_name** strings with the names of the moisture classes
**drying_model** if present,  must be 1 for each moisture class, the only value supported at present
**drying_lag** characteristic time (h) of approaching the FMC equilibrium from atmospheric conditions. If the atmospheric conditions were not changing, the difference between the FMC and the equilibrium would decrease by 1-exp(-1) = approx 63% in this time.
**wetting_model** if present, must be 1 for each moisture class, the only value supported at present, e.g., 1, 1, 1
**wetting_lag** the minimal characteristic time (h) of rain wetting, i.e., of approaching the saturated FMC, in very strong rain; the characteristic time of rain wetting decreases with rain intensity.  
**saturation_moisture** the saturated FMC after long rain,
**saturation_rain**  rain intensity (mm/h) at which the characteristic time of rain wetting attains 1/(1-exp(-1)) = approx 158% of its minimal value. 
**rain_threshold** rain intensity  (mm/h) ignored as too weak to wet anything
**fmc_gc_initialization**  initialization of FMC state for each FMC class, 0: from wrfinput, 1 = from fuelmc_g in namelist.input, 2 = from the FMC equilibrium with the atmosphere

LIST OF MODIFIED FILES: 
M       Registry/registry.fire
M       clean
M       dyn_em/module_initialize_fire.F
M       phys/module_fr_fire_core.F
M       phys/module_fr_fire_driver.F
M       phys/module_fr_fire_driver_wrf.F
M       phys/module_fr_fire_model.F
M       phys/module_fr_fire_phys.F
M       phys/module_fr_fire_util.F
M       test/em_fire/README.txt
M       test/em_fire/create_links.sh
A       test/em_fire/input_sounding_rain
A       test/em_fire/namelist.fire_fmc
A       test/em_fire/namelist.input_rain
M       test/em_real/namelist.input.fire


TESTS CONDUCTED: 
1. WTF from Github with modifications for Cheyenne, available in the WTF fork 
https://github.com/openwfm/WTF/tree/fergui (Pull request for WTF is pending). All tests passed except several that already fail in branch develop. 
2. Verified that the modification does not  change results including existing em_fire tests. 
3. Verified identical results in serial mode and with MPI on the test/em_fire/fmc and test/em_fire/rain fuel moisture test cases.

RELEASE NOTE: 
Fuel moisture content (FMC) model is coupled with the coupled fire-atmosphere model. FMC has a strong effect on fire spread, including the diurnal cycle, and previously was currently a constant in the fuel description.

Illustration of the effect of the coupled FMC model on diurnal fire spread variability in the 2012 Barker Canyon Fire, from the reference above:

![pages from corsica13-nhess](https://user-images.githubusercontent.com/6204560/53294946-ed342c00-37ad-11e9-8d0b-d50d268dc027.png)

davegill commented:
@janmandel @weiwangncar @dudhia @powersjg 
Jan,
A couple of points.

1. Small:
These are executables. Why would they be added?
``` 
test/em_fire/fmc/ideal.exe
test/em_fire/fmc/wrf.exe
```
These are existing binary files. These should be linked into the right directory during the build process.
```
A test/em_fire/rain/ETAMPNEW_DATA
A test/em_fire/rain/GENPARM.TBL
A test/em_fire/rain/LANDUSE.TBL
A test/em_fire/rain/RRTMG_LW_DATA
A test/em_fire/rain/RRTMG_SW_DATA
A test/em_fire/rain/SOILPARM.TBL
A test/em_fire/rain/URBPARM.TBL
A test/em_fire/rain/VEGPARM.TBL
```
If there are new links and new locations for executables, the "./clean" script needs to be updated. The links and executables probably need to be handled in the "./compile" script.

2. Bigger:
These changes are large enough to be considered either a new development or a significant enhancement. We have closed the release branch to new additions of this type. This PR will go onto the develop branch (as you have it). Since WRF is a public repository, anyone will be able to access this modification (when accepted, approximately the next few weeks, depending on testing and reviews). Being on the develop branch means that the code for this PR will be available on the master branch and as part of an official release the next major release (likely scheduled for April 2020).

davegill commented:
@janmandel 
Jan,
Please add a brief description in the comment section on how to activate the FMC (what namelist flags to set). Also, we do permit images in the comment section. If there is a striking "before vs after" figure that shows the impact of the fuel model on fire spread (etc), that would be great.

You should modify the WRF/run/README.namelist file to include your new settings. 

If appropriate, similar to the examples.namelist in the test/em_real directory, you may want to consider an annotated file in the em_fire directory to explain the myriad of options (it could also be named examples.namelist).

davegill commented:
@janmandel 
Jan,
Looking through your PR again, there were two more executables listed:
```
test/em_fire/rain/ideal.exe
test/em_fire/rain/wrf.exe
```
These would also be deleted from the PR, and the clean script would know to remove these as well.

janmandel commented:
Dave,

thank you very much for your quick review and suggestions. I understand
that we should have all links and test problem subdirectories in em_fire
created on compile and destroyed on clean -a but keep one set of soft links
in em_fire as the build system expects them to be there. There was already
script create_links.sh in the em_fire directory, which is executed from
compile em_fire. It used to do something similar but served no current
purpose, so we used that. I have now pushed an update with this change and
we are working on the rest.

Thanks,
Jan

On Tue, Feb 19, 2019 at 1:55 PM Dave Gill <notifications@github.com> wrote:

> @janmandel <https://github.com/janmandel>
> Jan,
> Looking through your PR again, there were two more executables listed:
>
> test/em_fire/rain/ideal.exe
> test/em_fire/rain/wrf.exe
>
> These would also be deleted from the PR, and the clean script would know
> to remove these as well.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/wrf-model/WRF/pull/792#issuecomment-465305858>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AF6skB2vyiuBHZ4KObhHjS3yLb6eL94Bks5vPGSpgaJpZM4a_4fr>
> .
>

adamk0 commented:
Two comments:
1. Even though we have 1h, 10h, 100h and 1000h fuel the model by default only uses the first three.
2. The initialization of the model is specified in tow places in namelist.input (fire_fmc_read) and in namelist.fire as (fmc_gc_initialization) I think it would be worthwhile to comment on that a little bit.

janmandel commented:
Dave,

We have updated the pull request with description with the new settings and a figure showing the effect of the new model, provided annotated namelist.input and namelist.fire in test/em_fire, made the test/em_fire subdirectories created on compile em_fire and deleted on clean -a, and added lines with the new settings to test/em_real/namelist.input.fire. We have also merged with the current develop branch and tested again. 

However, WRF/run/README.namelist does not contain any fire settings. I would prefer to address that in a separate pull request as it seems beyond the scope of this pull request concerning the fuel moisture model. It would take some time and a review of the various options and the registry variables in coordination with the Fire group.  For the same  reason, I have not attempted to update test/em_real/namelist.input.fire other than adding lines with the new settings. Please let me know if you want me do this as a part of this pull request anyway.

Thanks,

Jan

janmandel commented:
Ready for rewiew. Thanks! Jan

davegill commented on Registry/registry.fire:
```diff
@@ -67,6 +67,31 @@ state    real     burnt_area_dt *i*j      fire        1         z    hr    "BURN
+
+# fuel moisture model section
+dimspec  num_fmc    -   namelist=nfmc              z      fuel_moisture_classes 
+rconfig  integer  nfmc  namelist,fire  1    5  -   "nfmc" "number of fuel moisture classes"
```
@janmandel 
Jan,
Where are these new variables packaged? If not, they should be.

janmandel replied:
Dave,
all state variables in registry.fire are now packaged.
Jan

davegill commented on Registry/registry.fire:
```diff
@@ -8,7 +8,7 @@
 #         name>          namelist choice>          state vars>
 # 
-package   fire_sfire    ifire==2                    -             state:nfuel_cat,zsf,tign_g,rthfrten,rqvfrten,grnhfx,grnqfx,canhfx,canqfx,lfn,fuel_frac,fire_area,uf,vf,fgrnhfx,fgrnqfx,fcanhfx,fcanhfx,fcanqfx,ros,fxlong,fxlat,fuel_time,bbb,betafl,phiwc,r_0,fgip,ischap,lfn_hist,lfn_time,avg_fuel_frac,uah,vah
+package   fire_sfire    ifire==2                    -             state:nfuel_cat,zsf,tign_g,rthfrten,rqvfrten,grnhfx,grnqfx,canhfx,canqfx,lfn,fuel_frac,fire_area,uf,vf,fgrnhfx,fgrnqfx,fcanhfx,fcanhfx,fcanqfx,ros,fxlong,fxlat,fuel_time,bbb,betafl,phiwc,r_0,fgip,ischap,lfn_hist,lfn_time,avg_fuel_frac,uah,vah,fmc_g
```
@janmandel 
Jan,
This package statement looks like it was "clipped" off.

janmandel replied:
Dave,
I have now put all state variables in registry.fire in the package. 
Jan

davegill commented on phys/module_fr_fire_driver.F:
```diff
@@ -118,7 +127,68 @@ subroutine fire_driver_em ( grid , config_flags                    &
 
+    ! time - assume dt does not change
+    dt = grid%dt
+    time_start = itimestep * dt
```
@janmandel 
Jan,
What happens with a really small dt? Seems like time start could be zero for a while.

janmandel replied:
Dave,
this is fine, dt and time_start are both REAL, so from starting form itimestep=1, we have also time_start>0. These lines were there before, they were just moved up because time_start is needed earlier now.
Jan

davegill commented:
@janmandel 
> sf_sfclay_physics = 1 in &physics section

Jan,
1. Does this model _only_ work with the revised MM5 surface layer scheme?
2. There is a conflict in the clean script. Would you please address that.
3. Would you comment on reproducibility. For example, one run to the next, and also on differing numbers of processors.
4. Are you ready for us to have folks review the code?

janmandel commented:
Dave,

> Does this model only work with the revised MM5 surface layer scheme?

A surface model is needed to provide P2 Q2 PSFC. The initialization of the surface variables in ideal.exe (when built by ./compile em_fire) was tested only with  sf_sfclay_phys=1 and sf_surface_phys=1. I have clarified the comments in the pull request and in WRF/test/em_fire/namelist.input_rain.

> There is a conflict in the clean script. Would you please address that.

The develop branch changed in the meantime. We have now merged with the current develop at commit e883b53 , fixed the conflict, and rerun all tests.

> Would you comment on reproducibility. For example, one run to the next, and also on differing numbers of processors.

The results are identical from run to run, and as serial and with different numbers of MPI processes, tested by WTF and the new ideal test case WRF/test/em_fire/rain provided by the pull request. Issues that existed already in the baseline remain, which includes fire runs with OpenMP. I can look into those separately in future.

>Are you ready for us to have folks review the code?

Yes.

Thanks, 
Jan

janmandel commented:
@davegill @dudhia @domingom 
Hi, can you please consider this pull request - anything else we need to do? Thanks!
Jan

domingom commented:
Hi Jan, we are currently reviewing your pull request. We will get back to you with comments in the next few weeks.

domingom commented:
Jan,
 
We highly appreciate your effort in developing the fuel moisture model in WRF-Fire and committing the code to the repository. Having a fuel moisture model is a nice addition to the existing WRF-Fire capabilities. Here are some comments to make more user friendly the configuration of the fuel moisture model. Also some comments are oriented to define variables that have a physical meaning.

In general, we find more appropriate to replace the ‘fuel weights’ by fractions of the fuel load, which strictly depend on the fuel type. Otherwise, users can get confused with that nomenclature. Also, standard fuel models only have 1h, 10h, 100h, and live herbaceous fuel classes, so we would like to stick to that classification for consistency with the available datasets. Here is a list of more specific changes that will also make your additions more compatible with future extensions, including dynamic Scott & Burgan fuel model that we also plan to add to WRF’s next release.
 
-       Change fmc_gw0X arrays with X=[1,5], these are weights. These are proportion of fuel classes not moisture. Remove fmc_gw0X and change to actual fuel loads:
o   fmc_gw01 to fgi_1h
o   fmc_gw02 to fgi_10h
o   fmc_gw03 to fgi_100h
o   Remove fmc_gw04
o   fmc_gw05 to fgi_lh

-       Then, fuel weights are computed internally based on static fuel properties read in from the fire namelist.
-       Change &moisture to &fuel_moisture_model (there is moisture in the atmosphere as well). Actually, you can remove this namelist section, see comments below.
-       Remove moisture_classes entry from fire namelist (use the default of 4 classes: 1h, 10h, 100h and live herbaceous).
-       Remove drying_lag, since these are the established fuel classes (1h, 10, 100h, 1e9 for live herbaceous).
-       Remove saturation_rain and rain_threshold dependency on fuel class. Make them a single parameter. Perhaps, it would be better to have them be an internal parameter in the fuel moisture code.
-       Make wetting_lag a factor of the 10h fuel (only one entry in the namelist) and the rest of fuel classes are scaled internally in the code. Same for saturation_moisture (single value), defined as 1, not dimspec to nfmc size. In other words, remove num_fmc dependency.
-       Combine drying_model and wetting_model into a single namelist parameter (moisture_model), and also remove num_fmc dependency.
-       There is no need for a specific section on fuel moisture model in the namelist.fire. The user can turn on/off the fuel moisture model with a single namelist entry (recall comments above).
-       Replace fmc_gc by fmc_1h, fmc_10h, fmc_100h and fmc_lh (defined on the atmospheric grid). There should also be code added to module_initialize_real.F in order to ensure that the previous variables are properly read in from WPS met_* files (the METGRID.TBL needs to be aware of these 4 variables). If these variables are not present (at least the 10h fuel class), the code should stop (for fire_fmc_read=0) and print out a message indicating to switch to either fire_fmc_read=1 or fire_fmc_read=2. There should be code to initialize the remaining variables (e.g., fmc_1h=fmc_10h; fmc_100h=fmc_10h). Live herbaceous field is treated separately. If not present, then assumed to be all live (fmc_lh=1.20).
-       The modified fuel moisture (fmc_g(i,j)) is apparently only used to modify the resulting heat fluxes from the fire that fed back into the atmosphere. However, the fuel moisture content plays a major role in the rate of spread as well. You should include that contribution as well, which in fact will also affect the released heat fluxes through dmass. Therefore, each time the fuel moisture content is updated, there needs to be a call to ‘set_fire_params’ so a new default rate of spread (r_0(i,j)) is calculated, since it strongly depends on fuel moisture. Right now, this component seems to be missing.
-       Have you encountered any issues with the fuel moisture model for grid points at the fire front? We were wondering if these very high temperatures are going to induce a sudden dry out of the fuel and result in unrealistic behavior. It would perhaps be reasonable not to modify fuel moisture for points that are part of the active fire front, while still capturing the diurnal evolution of fmc prior to ignition.

Again, thanks for adding the fuel moisture model to WRF-Fire. Let me know if something is not clear or you need further feedback/clarifications regarding these comments.

janmandel commented:
@domingom @brankokosovic @pedro-jm

Hi Branko, Domingo, and Pedro, for some reason github made the request for review by itself once I pushed into the pull request branch. Not ready yet. I am working on the items in the summary below, which Adam sent us all after our meeting in July: 

We will keep the 5-layer fuel structure keep the 1000h load to maintain combability with the future fire models dealing with thermally thick 1000h fuel.
To keep the code flexible and in line with WRF conventions we will leave the num_fmc dependency defining the number of fuel classes, but we will remove it from the namelist template to prevent end users form accidental changes that could break the code.
To improve the transparency, the FMC_GC will be filled with the specific values from namelist called: fmc_1h, fmc_10h, fmc_100h, fmc_1000h and fmc_lh.
We will change the way how the fuel moisture is initialized, so it is added to met_em files and then processed by real.exe, not via geogrid how it is done right now.
Within the real.exe there will be a check if the fmc values are present in the met_em files, if not, the fmc_XX values form the namelist will be populated into FMC_GC. Pedro will send an example how to do that.
The initialization from the equilibrium will be activated by a separate switch in the namelist, and the activation will be executed from the fire code, overwriting whatever fmc_gc was in the wrfinput file.
We will rename the moisture section and consolidate drying / wetting lags.
Most of the advanced functions will be removed from the template namelist in order to simplify it, but still the ability to override the default values by namelist entries will be kept in place.
We will fix the linkage between the ROS and the fuel moisture that got broken during transplanting the fuel moisture code.
We will work on the adding the fuel moisture variable description to the namelist.input_readme while Branko, Pedro and Domingo will update the rest of the fire-related variables.
We don’t expect any problems with unphysical drying of the fuel near the fireline (long drying lag, and the dilution of the heat flux in larger atmospheric cells).

davegill commented:
@pedro-jm 
Pedro,
Would you please provide a review of this PR. We are trying to finalize the contents for version 4.2.

pedro-jm commented:
@davegill 
OK, we will take it a look and send feedback.

janmandel commented:
Thank you. Testing is in progress. I do not expect significant changes at this point. I plan to post a summary of new developments since July soon.

janmandel commented:
@pedro-jm @domingom @brankokosovic

Thank you for your valuable comments, which helped to improve the code. Here is a summary of the new changes. Further details can be found in https://github.com/wrf-model/WRF/pull/792#issuecomment-544225413.

namelist.fire is now optional and serves only to override the statically coded defaults. File test/em_fire/namelist.fire_fmc contains values identical to the defaults. The rain test case gives the same results with or without namelist.fire.

We have added in namelist.fire the option fmc_gc_initialization(k) = 3 to initialize FMC_GC from scalars named fmc_1h, fmc_10h, fmc_100h, fmc_1000h, fmc_live. The default is initialization from the equilibrium for dead fuel, and constant 100% for live fuel. This corresponds in namelist.fire to
fmc_gc_initialization=    2, 2,  2,  2,  3,
fmc_live = 1.0,

Input of weights fmc_gw* in each fuel category was replaced by fuel loads fgi_1h, fgi_10h, fgi_100h, fgi_1000h, fgi_live. Since we no longer input weights that add up to one, there us the issue in which units are fgi_* given, and we are responsible for checking that (fgi_1h+fgi_10h+fgi_100h+fgi_1000h+fgi_live)/fgi=1 
in each fuel category. We have coded defaults for fgi_* from a table in Anderson 1982 (tons/acre) to enable a quick visual check against literature, while existing fgi is coded as kg/m^2. Thus the ratios are not one but they are the same in all fuel categories, up to minor variations in the table entries from the literature.

We have kept the internal code structure general,  using arrays with the index k=1:num_fmc (currently, num_fmc=5) instead of individual slices,  and with the models and their coefficients all statically coded but changeable from namelist.fire. This is important for the support of models fitted to data and for extensions, e.g., including duff for FCCS. 
 
To make the changes in module_initialize_real.F to set defaults, we will need some help, and we will have to develop a pathway through metgrid hand together with a test case. In particular, we will have to start from a working real fire case with WRF4, which we do not have. Can you please provide one? Also, this seems outside of the scope of this pull request, which was to  port the fuel moisture model from https://github.com/openwfm/wrf-fire/master. We would be happy to work on a collaborative pull request for this real.exe feature in future.

We have now ported also the use of extended model parameters fields fmep from wrfinput; currently there are two (i.e, k=1:2), namely spatially variable adjustments to the equilibrium moisture and the saturation moisture. This feature (1460883bff38a673a29facdaa1c6aa19c7b38) was omitted in the initial port but it is essential for the use of fuel moisture fields from RAWS data assimilation (c1236aabc83592651ddebf148be93efcc46d46ad) or a similar system in future. If the fmep feature is unused, fmep is zero, and nothing changes.

We have fixed the linkage between the ROS and the fuel moisture that got broken during initial porting process.

WTF tests passed as in the original pull request.

At the moment, this pull request is based on branch develop at release 4.1, since the new km_opt=5 feature in the current branch develop at d059afe1e646f772c010c48a5eb630ebd8f82d73 breaks the test problems in test/em_fire. After some efforts, we have opted to roll back and rebuild. After a complete fix for d059afe1e646f772c010c48a5eb630ebd8f82d73 is merged into branch develop (this should be https://github.com/wrf-model/WRF/pull/1026), we will merge with branch develop again and retest.

davegill commented:
@janmandel 
1. Thanks for the informative update to recap all of the work. It is very helpful.
2. We have a temporary fix in place in the develop branch that should allow your code to run. There is a permanent fix also to the develop branch for this km_opt=5 feature, but that has not been merged yet.

davegill commented:
@janmandel 
Jan,
Small things now - 
Would you please update the PR commit message. There should be 15 files modified, and you only list 13.

davegill commented:
@janmandel 
Jan,
The km_opt=5 bug should be fixed. It is in the develop branch. Locally in a sandbox (not in this PR), you can do a test by merging in the new develop branch. After you are happy, let me know.

domingom commented:
Hi Jan,

Here are the final comments to your PR.

We really appreciate all the code modifications you have made to
accommodate our previous comments/suggestions. Here a few additional
comments, which should not require any major code changes.



-       Add a new package for the fuel moisture model (FMM) variables,
similar to ifire==2, using fmoist_run variable. This will reduce the
allocated memory when the FMM is not used.

-       Perhaps remove fmoist_freq and just leave fmoist_dt to avoid
confusion.

-       Maybe not now but for the future consider not doing interpolation
from atmospheric to fire grid fuel moisture rather downscaling (i.e., all
of the values within an atmospheric grid cell will have the same fuel
moisture values of each class). This would then lead to removal of halo
exchange for fmc_g and fmc_gc.

-       In module_fr_fire_driver.F, there is a comment about “time – assume
dt does not change”. Is that true or does the FFM work also with time
varying dt?

-       In general, does the FFM work with restart? Are the logical
variables such as run_advance_moisture, moisture_time, run_fuel_moisture,
moisture_initializing properly set when restarting?

-        The call to fire_driver_em() has now fire_ifun_end = 2, where it
used to be 3. According to the code comments in module_fr_fire_driver.F,
fire_ifun_end = 3 has to do with initialization of time step. Is that
skipped now or embedded within fire_ifun_end = 2?

-       How are the fuel classes initialized for a real case? It seems that
the FM values are prescribed from the solution of the fuel moisture model?
Could you document/explain that?



To ensure backward compatibility, we would like to perform a test to ensure
that switching off the FMM does produce the same results as the existing
WRF-Fire code. We will wait until we hear from you about these latest
comments. After the backward compatibility check the code would ready to be
directly merged into WRF’s next release.



Thanks!



Domingo, Branko and Pedro

----------------------------------------------------------------
Dr. Domingo Muñoz-Esparza
Project Scientist II
Research Applications Laboratory
National Center for Atmospheric Research
P.O. Box 3000
Boulder, CO 80307-3000
phone: (303) 497-8384
e-mail: domingom@ucar.edu
webpage: *https://staff.ucar.edu/users/domingom
<https://staff.ucar.edu/users/domingom>*


On Mon, Jan 27, 2020 at 6:23 PM Dave Gill <notifications@github.com> wrote:

> @janmandel <https://github.com/janmandel>
> Jan,
> The km_opt=5 bug should be fixed. It is in the develop branch. Locally in
> a sandbox (not in this PR), you can do a test by merging in the new develop
> branch. After you are happy, let me know.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/wrf-model/WRF/pull/792?email_source=notifications&email_token=AETGQTQPEJXSXRRHTNHEFFLQ76CKPA5CNFSM4GX7Q7V2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEKBWEGY#issuecomment-579035675>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AETGQTUQMOFISD6VWADSYPDQ76CKPANCNFSM4GX7Q7VQ>
> .
>

janmandel commented:
> @janmandel
> Jan,
> The km_opt=5 bug should be fixed. It is in the develop branch. Locally in a sandbox (not in this PR), you can do a test by merging in the new develop branch. After you are happy, let me know.

Hi Dave,
we have merged with develop on Dec 7 which temporarily fixed the issue for us. I see new  dd0c2094fc16ed1106 in develop as of today, we'll merge with develop again and retest.
Thanks, Jan

davegill commented:
@janmandel 
Jan,

>To ensure backward compatibility, we would like to perform a test to ensure
that switching off the FMM does produce the same results as the existing
WRF-Fire code. We will wait until we hear from you about these latest
comments. After the backward compatibility check the code would ready to be
directly merged into WRF’s next release.

This test has been requested prior to inclusion in the source code. We are pushing to release this week, so time is pretty critical for this testing. During our next phase (through early March), we will get the rest of the small pieces handled.

janmandel commented:
> @janmandel
> Jan,
> 
> > To ensure backward compatibility, we would like to perform a test to ensure
> > that switching off the FMM does produce the same results as the existing
> > WRF-Fire code. We will wait until we hear from you about these latest
> > comments. After the backward compatibility check the code would ready to be
> > directly merged into WRF’s next release.
> 
> This test has been requested prior to inclusion in the source code. We are pushing to release this week, so time is pretty critical for this testing. During our next phase (through early March), we will get the rest of the small pieces handled.

Hi Dave

OK, I will prioritize this test. I already checked testing before and after merge with current develop few days ago and all was well - I wanted to run some more to be sure before anwering that. 

Just to make clear: you want me to compare the pull request branch and  the develop branch on fire examples with the same input files. I will push this ahead of the other things on the list. 

Thanks Jan

domingom commented:
HI Jan,

Yes. In particular, when the new fuel model developments are off. Just a
sanity check to make sure nothing else got accidentally modified.

Thanks!

Domingo

----------------------------------------------------------------
Dr. Domingo Muñoz-Esparza
Project Scientist II
Research Applications Laboratory
National Center for Atmospheric Research
P.O. Box 3000
Boulder, CO 80307-3000
phone: (303) 497-8384
e-mail: domingom@ucar.edu
webpage: *https://staff.ucar.edu/users/domingom
<https://staff.ucar.edu/users/domingom>*


On Tue, Feb 4, 2020 at 12:16 PM Jan Mandel <notifications@github.com> wrote:

> @janmandel <https://github.com/janmandel>
> Jan,
>
> To ensure backward compatibility, we would like to perform a test to ensure
> that switching off the FMM does produce the same results as the existing
> WRF-Fire code. We will wait until we hear from you about these latest
> comments. After the backward compatibility check the code would ready to be
> directly merged into WRF’s next release.
>
> This test has been requested prior to inclusion in the source code. We are
> pushing to release this week, so time is pretty critical for this testing.
> During our next phase (through early March), we will get the rest of the
> small pieces handled.
>
> Hi Dave
>
> OK, I will prioritize this test. I already checked testing before and
> after merge with current develop few days ago and all was well - I wanted
> to run some more to be sure before anwering that.
>
> Just to make clear: you want me to compare the pull request branch and the
> develop branch on fire examples with the same input files. I will push this
> ahead of the other things on the list.
>
> Thanks Jan
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/wrf-model/WRF/pull/792?email_source=notifications&email_token=AETGQTUT7P6AGIS3ADZLDGDRBG5JVA5CNFSM4GX7Q7V2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEKY3ACY#issuecomment-582070283>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AETGQTUL7GVYUXOHFGVCG2LRBG5JVANCNFSM4GX7Q7VQ>
> .
>

janmandel commented:
> HI Jan, Yes. In particular, when the new fuel model developments are off. Just a sanity check to make sure nothing else got accidentally modified. Thanks! Domingo
> […](#)

Hi Domingo

yes when using the same input files as for the existing code in the develop branch now the fuel moisture model should be off by default and the output should be the same.

Jan

davegill commented:
@janmandel 
Jan,
Since we are trying hard to make a deadline, I just want to be VERY clear:

>yes when using the same input files as for the existing code in the develop branch now the fuel moisture model should be off by default and the output should be the same.

This is the test to conduct. I understand that saying "should" is a polite way to provide expected information before the test is complete. However, since we are up against the deadline, I do not want anyone surprised about additional tests that need to be done.

janmandel commented:
> @janmandel
> Jan,
> Since we are trying hard to make a deadline, I just want to be VERY clear:
> 
> > yes when using the same input files as for the existing code in the develop branch now the fuel moisture model should be off by default and the output should be the same.
> 
> This is the test to conduct. I understand that saying "should" is a polite way to provide expected information before the test is complete. However, since we are up against the deadline, I do not want anyone surprised about additional tests that need to be done.

Done, all good. Branch fuel_moisture_model at 46b548370c9eff7c9a8 and branch develop at 33bf83c664426b267452 give the same wrfout files, except XLONG and XLAT which are now set in ideal.exe and the fuel moisture variables, which do not exist in branch develop, and are not in registry packages yet. Please see https://github.com/openwfm/WRF-Fire-merge/wiki/Fuel-moisture-model-pull-request-final-changes-2020 for details.

davegill commented:
@domingom @pedro-jm @brankokosovic @dudhia @weiwangncar @janmandel 
Domingo, Pedro, Branko,
This PR looks to be ready for a review. There are a few small items, but we should be able to get to those in the next couple of weeks. We need the WRF Fire team to give an official approval in a review for this to move forward.

janmandel commented:
Updated the PR comment by new list of changed files from git diff --name-status develop

janmandel commented:
@davegill @domingom @pedro-jm @brankokosovic @dudhia @weiwangncar

Please see summary maintained at https://github.com/openwfm/WRF-Fire-merge/wiki/Fuel-moisture-model-pull-request-final-changes-2020 .

janmandel commented:
@domingom @pedro-jm @brankokosovic
Tested registry packages, which had been added in d564d7da21e33eaec9fd3ef235cb74a8ec48edd5, to save memory when fmoist_run=.false. and/or fmoist_interp=.false. (default with existing namelist.input).

janmandel commented:
@domingom @pedro-jm @brankokosovic @davegill

_In general, does the FFM work with restart? Are the logical
variables such as run_advance_moisture, moisture_time, run_fuel_moisture,
moisture_initializing properly set when restarting?_

Yes, we use restart routinely. @Fergui now tested restart in the ported code in this pull request on examples with and without the fuel moisture model to be sure, and the outputs are the same with and without restart. See https://github.com/openwfm/WRF-Fire-merge/wiki/Fuel-moisture-model-pull-request-final-changes-2020#does-the-ffm-work-with-restart-done for details.

janmandel commented:
@domingom @pedro-jm @brankokosovic @davegill 

Here are the few remaining items. I appreciate your constructive comments which helped improve the PR.

_Perhaps remove fmoist_freq and just leave fmoist_dt_

Good idea; we use fmoist_dt mostly anyway. However, this close to the release I would prefer to make only truly necessary changes; any change at this point makes it necessary to repeat the testing. Perhaps in future then.

_Maybe not now but for the future consider not doing interpolation from atmospheric to fire grid fuel moisture rather downscaling (i.e., all of the values within an atmospheric grid cell will have the same fuel moisture values of each class). This would then lead to removal of halo exchange for fmc_g and fmc_gc._

Maybe as an option in future then, e.g., we should test if this does not introduce visible blockiness at resolutions we use.

_|n module_fr_fire_driver.F, there is a comment about “time – assume dt does not change”. Is that true or does the FFM work also with time varying dt?_

Currently the time keeping supports fixed time step only.

_The call to fire_driver_em() has now fire_ifun_end = 2, where it used to be 3. According to the code comments in module_fr_fire_driver.F, fire_ifun_end = 3 has to do with initialization of time step. Is that skipped now or embedded within fire_ifun_end = 2?_

This comment refers to the line https://github.com/openwfm/WRF-Fire-merge/blob/d564d7da21e33eaec9fd3ef235cb74a8ec48edd5/phys/module_fr_fire_driver_wrf.F#L51 , where in subroutine fire_driver_em_init, fire_driver_em is called with the argument fire_ifun_end = 2.

By design, initialization is done in passes ifun=1 to 2, while each time step consists of passes ifun=3 to ifun=6. The correct value is 3. I changed fire_ifun_end = 2 in the initialization to 3 in 73fcb8565fa104c on Sep 6 2010 to have fire winds early on for diagnostics. The code included in WRF in cd867781f4 on Dec 30 2010 still had 3. I changed 3 back to 2 later in 2f4650eecd0e9dc4a9 on Dec 15 2012 but at that point we could no longer update the fire code in WRF. The mergediff tool noted the difference and I picked the original 2, which did not change results as verified in the [tests](https://github.com/wrf-model/WRF/pull/792#issuecomment-582203707) @davegill asked for.

_How are the fuel classes initialized for a real case? It seems that the FM values are prescribed from the solution of the fuel moisture model? Could you document/explain that?_

The variables FMC_GC and FMEP are assumed by WRF to be set in wrfinput. This is done in WPS, outside of the scope of this PR. We set them by running a version of the model in Python with [RAWS data assimilation]( https://github.com/openwfm/wrfxpy/blob/master/rtma_cycler.sh ). This is at the moment done [through GEOGRID]( https://github.com/openwfm/wrfxpy/blob/master/tests/fmda_geogrid.json). As we discussed last summer, METGRID is more appropriate for that purpose. It would also allow us to input several time levels of FMC_GC and FMEP so that is definitely the plan. The FMEP variable contains spatially variable adjustments to model parameters from RAWS data assimilation.
